File: VehicleLocations.py
import requests


class VehicleLocations:

    def obtain_json_vehicles(self):
        mbta_url = "https://api-v3.mbta.com/vehicles?filter[route]=" \
                   "Red," \
                   "Blue," \
                   "Orange," \
                   "Green-B," \
                   "Green-C," \
                   "Green-D," \
                   "Green-E"

        response = requests.get(mbta_url)

        if response.status_code == 200:
            return response.json()['data']

    def process_vehicle_json(self, vehicle_json):
        vehicle_list = []

        for i in range(len(vehicle_json)):
            vehicle_list.append([vehicle_json[i]['relationships']['route']['data']['id'],
                                 vehicle_json[i]['attributes']['latitude'],
                                 vehicle_json[i]['attributes']['longitude']])

        return vehicle_list

    def return_vehicle_locations(self):
        vehicle_json = self.obtain_json_vehicles()
        return self.process_vehicle_json(vehicle_json)


if __name__ == '__main__':
    VehicleLocations().return_vehicle_locations()

File: src/vehicle_locations.cpp
#include "vehicle_locations.h"

VehicleLocations::VehicleLocations(){};

Json::Value VehicleLocations::obtainJsonVehicles() {
    const int OK_RESPONSE = 200;

    // 0 = green line, 1 = subway
    const std::string MBTA_URL =
        "https://api-v3.mbta.com/vehicles?filter[route]="
        "Red,"
        "Blue,"
        "Orange,"
        "Green-B,"
        "Green-C,"
        "Green-D,"
        "Green-E";

    cpr::Response r = cpr::Get(cpr::Url{MBTA_URL});

    Json::Reader reader;
    Json::Value root;

    if (r.status_code == OK_RESPONSE) {
        bool parseSuccessful = reader.parse(r.text, root);

        if (parseSuccessful) {
            return root["data"];
        }
    }
};

std::list<Vehicle>
VehicleLocations::getListVehicles(Json::Value& jsonVehicles) {

    std::list<Vehicle> listVehicles;

    for (int i = 0; i < jsonVehicles.size(); i++) {
        Vehicle vehicle(
            jsonVehicles[i]['relationships']['route']['data']['id'].asString(),
            jsonVehicles[i]["attributes"]["latitude"].asString(),
            jsonVehicles[i]["attributes"]["longitude"].asString());

        listVehicles.push_front(vehicle);
    };

    return listVehicles;
};

std::list<Vehicle> VehicleLocations::returnJsonVehicles() {
    Json::Value vehiclesJson = obtainJsonVehicles();
    return getListVehicles(vehiclesJson);
};
